File: MST/Kruskal.cpp
#include <vector>
#include <algorithm>
#include <iostream>
#include <tuple>
using namespace std;

class UnionFind {
public:
    vector<int> parent, rank;
    UnionFind(int size) : parent(size), rank(size, 0) {
        for (int i = 0; i < size; ++i) parent[i] = i;
    }

    int find(int x) {
        if (x != parent[x]) parent[x] = find(parent[x]);
        return parent[x];
    }

    void unite(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX != rootY) {
            if (rank[rootX] < rank[rootY]) swap(rootX, rootY);
            parent[rootY] = rootX;
            if (rank[rootX] == rank[rootY]) rank[rootX]++;
        }
    }
};

class KruskalMST {
public:
    int V; // Number of vertices
    vector<tuple<int, int, int>> edges; // (weight, start, end)

    KruskalMST(int V) : V(V) {}

    void addEdge(int u, int v, int weight) {
        edges.push_back(make_tuple(weight, u, v));
    }

    vector<tuple<int, int, int>> constructMST() {
        sort(edges.begin(), edges.end());
        UnionFind uf(V);
        vector<tuple<int, int, int>> result;

        for (auto &[weight, u, v] : edges) {
            if (uf.find(u) != uf.find(v)) {
                uf.unite(u, v);
                result.push_back({u, v, weight});
            }
        }
        return result;
    }
};

int main() {
    int V = 4; // 정점의 수
    KruskalMST kruskal(V);
    
    // 간선 추가: u, v, 가중치
    kruskal.addEdge(0, 1, 10);
    kruskal.addEdge(0, 2, 6);
    kruskal.addEdge(0, 3, 5);
    kruskal.addEdge(1, 3, 15);
    kruskal.addEdge(2, 3, 4);
    
    auto mst = kruskal.constructMST();

    cout << "Edges in the MST:" << endl;
    for (auto &[u, v, weight] : mst) {
        cout << u << " - " << v << " : " << weight << endl;
    }

    return 0;
}

File: MST/Kruskal.py
class UnionFind:
    def __init__(self, size):
        self.root = [i for i in range(size)]
        self.rank = [1] * size

    def find(self, x):
        if x == self.root[x]:
            return x
        self.root[x] = self.find(self.root[x])  # Path compression
        return self.root[x]

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.root[rootY] = rootX
            elif self.rank[rootX] < self.rank[rootY]:
                self.root[rootX] = rootY
            else:
                self.root[rootY] = rootX
                self.rank[rootX] += 1

def kruskal(graph_edges, n):
    uf = UnionFind(n)
    mst = []
    cost = 0

    graph_edges.sort(key=lambda x: x[2])  # Sort edges by weight

    for edge in graph_edges:
        u, v, weight = edge
        if uf.find(u) != uf.find(v):
            uf.union(u, v)
            mst.append((u, v, weight))
            cost += weight

    return mst, cost

# 예제 사용
edges = [(0, 1, 10), (0, 2, 6), (0, 3, 5), (1, 3, 15), (2, 3, 4)]
n = 4  # 정점의 수
mst, total_cost = kruskal(edges, n)
print("MST:", mst)
print("Total Cost:", total_cost)
